dot/parachain/backing/statement_table.go

```diff
@@ -113,25 +121,246 @@ func (table *statementTable) getCommittedCandidateReceipt(candidateHash parachai
+		oldCandidateHash := proposals[0].candidateHash
+		oldSignature := proposals[0].signature
+
+		// if digest is different, fetch candidate and note misbehaviour.
+		if oldCandidateHash != candidateHash {
+			data, ok := table.candidateVotes[oldCandidateHash]
+			if !ok {
+				// when proposal first received from authority, candidate votes entry is created.
+				// and here proposals is not empty, so candidate votes entry should be present.
+				// So, this should never happen.
+				panic(fmt.Sprintf("%s for candidate-hash: %s", errCandidateDataNotFound, oldCandidateHash))
+			}
+
+			oldCandidate := data.candidate
+
+			misbehaviour := parachaintypes.MultipleCandidates{
+				First: parachaintypes.CommittedCandidateReceiptAndSign{
+					CommittedCandidateReceipt: oldCandidate,
+					Signature:                 oldSignature,
+				},
+				Second: parachaintypes.CommittedCandidateReceiptAndSign{
+					CommittedCandidateReceipt: candidate,
+					Signature:                 signature,
+				},
+			}
+			return nil, misbehaviour, nil
+		}
+	case table.config.allowMultipleSeconded && isCandidateAlreadyProposed(proposals, candidateHash):
+		// nothing to do here.
+	default:
+		proposals = append(proposals, proposal{candidateHash, signature})
+		table.authorityData[authority] = proposals
+
+		table.addCandidateVote(candidateHash, paraID, candidate)
+	}
+
+	return table.validityVote(authority, candidateHash,
+		validityVoteWithSign{validityVote: issued, signature: signature}, tableCtx)
+}
+
+func (table *statementTable) addCandidateVote(
+	candidateHash parachaintypes.CandidateHash,
+	paraID parachaintypes.ParaID,
+	candidate parachaintypes.CommittedCandidateReceipt,
+) {
+	table.candidateVotes[candidateHash] = &candidateData{
+		groupID:       paraID,
+		candidate:     candidate,
+		validityVotes: make(map[parachaintypes.ValidatorIndex]validityVoteWithSign),
+	}
+}
+
+func (table *statementTable) validityVote(
+	from parachaintypes.ValidatorIndex,
+	candidateHash parachaintypes.CandidateHash,
+	voteWithSign validityVoteWithSign,
+	tableCtx *tableContext,
+) (*Summary, parachaintypes.Misbehaviour, error) {
+	data, ok := table.candidateVotes[candidateHash]
+	if !ok {
+		return nil, nil, errCandidateDataNotFound
+	}
+
+	// check that this authority actually can vote in this group.
+	if !tableCtx.isMemberOf(from, data.groupID) {
+		switch voteWithSign.validityVote {
+		case valid:
+			validStatement := parachaintypes.NewStatementVDT()
+			err := validStatement.Set(parachaintypes.Valid(candidateHash))
+			if err != nil {
+				return nil, nil, fmt.Errorf("setting valid statement: %w", err)
+			}
+
+			misbehaviour := parachaintypes.UnauthorizedStatement{
+				Payload:        validStatement,
+				ValidatorIndex: from,
+				Signature:      voteWithSign.signature,
+			}
+
+			return nil, misbehaviour, nil
+		case issued:
+			panic("implicit issuance vote must only cast from `importCandidate` after checking group membership of issuer.")
```

Comment:
What could lead to this case?  I'm trying to understand why this isn't some sort of misbaviour, but is something that would cause a panic instead?

Reply:
the panic message explains it, but still, for more context,

in the `importStatement` method, in the case of 
- **valid vote**(Valid statement), We call the **`validityVote` method**.
- **implicit issuance vote**(Seconded statement), we call **importCandidate method**.
    - in the importCandidate method, we check if the validator is a member of the group of parachain validators. if
        - **it is not a member; it is misbehaviour**
        - otherwise, proceed further, and  if a validator does not already propose the candidate, We call the **`validityVote` method**
        
As we already ensured the validator's membership in the `importCandidate` method, the `validityVote` method should not get the case where the validator is not a member of the parachain validators.

If we get this case, it means implementation is incorrect.

Reply:
This is helpful, thanks for explaining.